File: src/entropy_redundancy.py
import math
from collections import Counter

def calculate_entropy(text):
    if not text:
        return 0.0
    
    freq = Counter(text)
    entropy = 0.0
    for count in freq.values():
        p = count / len(text)
        entropy -= p * math.log2(p)
    
    return entropy

def calculate_redundancy(text, alphabet_size=256):
    if not text or alphabet_size <= 1:
        return 0.0
    
    entropy = calculate_entropy(text)
    max_entropy = math.log2(alphabet_size)
    redundancy = max_entropy - entropy
    
    return max(redundancy, 0.0)

def main():
    input_str = input("Enter a string of characters: ")
    
    entropy = calculate_entropy(input_str)
    redundancy = calculate_redundancy(input_str)
    
    print(f"Entropy: {entropy}")
    print(f"Redundancy: {redundancy}")

if __name__ == "__main__":
    main()

File: src/entropy_redundancy.cpp
#include <iostream>
#include <iomanip>
#include <string>
#include <map>
#include <cmath>

using namespace std;

double calculate_entropy(const string &text) {
    if (text.empty()) {
        return 0.0;
    }

    map<int, int> freq;
    for (char c : text) {
        freq[static_cast<unsigned char>(c)]++;
    }

    double entropy = 0.0;
    for (const auto &pair : freq) {
        double p = static_cast<double>(pair.second) / text.size();
        entropy -= p * (log(p) / log(2.0));
    }
    return entropy;
}

double calculate_redundancy(const string &text, int alphabet_size = 256) {
    if (text.empty() || alphabet_size <= 1) {
        return 0.0;
    }

    double entropy = calculate_entropy(text);
    double max_entropy = log(static_cast<double>(alphabet_size)) / log(2.0);
    double redundancy = max_entropy - entropy;
    return redundancy >= 0.0 ? redundancy : 0.0;
}

int main() {
    string input;
    cout << "Enter a string of characters: ";
    getline(cin, input);

    double entropy = calculate_entropy(input);
    double redundancy = calculate_redundancy(input, 256);

    cout << fixed << setprecision(4);
    cout << "Entropy: " << entropy << '\n';
    cout << "Redundancy: " << redundancy << '\n';
    return 0;
}
